src/chainstate/burn/operations/leader_key_register.rs

```diff
@@ -108,89 +133,89 @@ where
             return Err(op_error::InvalidInput);
         }
 
-        if tx.outputs.len() < 1 {
-            test_debug!("Invalid tx: inputs: {}, outputs: {}", tx.inputs.len(), tx.outputs.len());
+        if outputs.len() < 1 {
+            test_debug!("Invalid tx: inputs: {}, outputs: {}", inputs.len(), outputs.len());
             return Err(op_error::InvalidInput);
         }
 
-        if tx.opcode != OPCODE {
-            test_debug!("Invalid tx: invalid opcode {}", tx.opcode);
+        if tx.opcode() != Opcodes::LeaderKeyRegister as u8 {
+            test_debug!("Invalid tx: invalid opcode {}", tx.opcode());
             return Err(op_error::InvalidInput);
         }
 
-        let parse_data_opt = LeaderKeyRegisterOp::<A, K>::parse_data(&tx.data);
-        if parse_data_opt.is_none() {
-            test_debug!("Invalid tx data");
-            return Err(op_error::ParseError);
-        }
+        let data = match LeaderKeyRegisterOp::parse_data(&tx.data()) {
+            Some(data) => {
+                data
+            },
+            None => {
+                test_debug!("Invalid tx data");
+                return Err(op_error::ParseError);
+            }
+        };
 
-        let data = parse_data_opt.unwrap();
-        let address = tx.outputs[0].address.clone();
+        let address = outputs[0].address.clone();
 
         Ok(LeaderKeyRegisterOp {
             consensus_hash: data.consensus_hash,
             public_key: data.public_key,
             memo: data.memo,
             address: address,
 
-            op: OPCODE,
-            txid: tx.txid.clone(),
-            vtxindex: tx.vtxindex,
-            block_number: block_height,
+            txid: tx.txid(),
+            vtxindex: tx.vtxindex(),
+            block_height: block_height,
             burn_header_hash: block_hash.clone(),
-
-            _phantom: PhantomData
         })
     }
 }
 
-impl<A, K> BlockstackOperation<A, K> for LeaderKeyRegisterOp<A, K> 
-where
-    A: Address,
-    K: PublicKey
-{
-    fn from_tx(block_height: u64, block_hash: &BurnchainHeaderHash, tx: &BurnchainTransaction<A, K>) -> Result<LeaderKeyRegisterOp<A, K>, op_error> {
-        LeaderKeyRegisterOp::<A, K>::parse_from_tx(block_height, block_hash, tx)
+impl BlockstackOperation for LeaderKeyRegisterOp {
+    fn from_tx(block_header: &BurnchainBlockHeader, tx: &BurnchainTransaction) -> Result<LeaderKeyRegisterOp, op_error> {
+        LeaderKeyRegisterOp::parse_from_tx(block_header.block_height, &block_header.block_hash, tx)
     }
 
-    fn check(&self, burnchain: &Burnchain, conn: &DBConn) -> Result<CheckResult, op_error> {
+    fn check<'a>(&self, burnchain: &Burnchain, block_header: &BurnchainBlockHeader, tx: &mut BurnDBTx<'a>) -> Result<(), op_error> {
+        // this will be the chain tip we're building on
+        let chain_tip = BurnDB::get_block_snapshot(tx, &block_header.parent_block_hash)
+            .expect("FATAL: failed to query parent block snapshot")
```

Comment:
Should this panic? If a leader submits a bad block header with a bad parent hash, that should just get ignored by network nodes -- or is that checked previously in the block processing, and if the code reaches this point, it should panic?

Reply:
`block_header.parent_block_hash` is the Bitcoin block's parent block hash, not the Stacks block's parent block hash.  The `block_header` value comes from data fetched by `Burnchain::sync`, which pulls block headers and blocks from Bitcoin.  Bitcoin blocks are processed in hash-chain order, so a parent block's snapshot should always be calculated and stored to the `BurnDB` before any of its children.  If `BurnDB::get_block_snapshot` fails to load a given Bitcoin block's parent block's snapshot, it indicates a serious (unrecoverable) problem.